google-cloud-firestore/src/main/java/com/google/cloud/firestore/UpdateBuilder.java

```diff
@@ -723,40 +735,88 @@ BatchState getState() {
     return state;
   }
 
-  Set<DocumentReference> getDocuments() {
-    return documents;
+  boolean hasPath(String path) {
+    return this.pendingOperations.containsKey(path);
+  }
+
+  Set<String> getPendingDocs() {
+    return this.pendingOperations.keySet();
+  }
+
+  void markComplete() {
+    completeFuture.set(null);
   }
 
-  int getOperationCount() {
-    return resultsMap.size();
+  int getPendingOperationCount() {
+    return pendingOperations.size();
+  }
+
+  /**
+   * Removes all operations not specified in documentsToRetry from the batch. Marks the batch as
+   * READY_TO_SEND in order to allow the next batch to be sent.
+   */
+  void sliceBatchForRetry(final Set<String> documentsToRetry) {
```

Comment:
Oh man. I really tried to use the BulkCommitBatch constructor directly to properly imitate Node. However, since the `BulkCommitBatch` logic lives inside `UpdateBuilder`, it got pretty ugly. I had to add a bunch of getters, and managing the state on a `BulkCommitBatch`  becomes a tricky problem / potential for future bugs. I think slicing the batch here is the lesser of the two evils, and I'm open to separating out the batch state override.

Reply:
How about https://gist.github.com/schmidt-sebastian/48ed97084b74bf7e3dd5936eddbfc001, or is this only addressing half of the problem?

Reply:
I ran into two main issues: first, my currently implementation requires the batch in `bulkCommit()` to be marked `final`, which means it can't be changed directly, though I guess we could put it in an array. The second is that the `completeFuture` used to track `awaitBulkCommit()` will never resolve, unless we pipe it through somehow (is that possible?).

Reply:
Update: Ended up updating to use BulkCommitBatch as constructor instead of slicing, pending `successfulAsList`